fill in LIP3 (#56)
libra/lip: lips/lip-3.md

> created: 06/09/2020

Reviewer: This is a great doc.  I would like it to live in the Libra core repository so it can be updated with minor changes (e.g. reformatting/typo fixes, more detail) outside of the LIP process.  It seems like a good fit for Libra specifications (https://github.com/libra/libra/tree/master/specifications), thoughts?

---

# Introduction

> Libra Upgrade Management (LUM) is a set of protocols for the Libra Networks company to reconfigure the blockchain. With respect to time-sensitive decisions and incident response, LUM provides the Libra Networks--a subsidiary of the Libra Association licensed in Switzerland to operate a payment system--with technical means to respond to critical incidents, as (i) necessitated by its license terms and the law and (ii) as determined by the Libra Network risk appetite/tolerance policy.

Reviewer: ```suggestion
Libra Upgrade Management (LUM) is a set of protocols for the Libra Networks company (LN) to reconfigure the blockchain. With respect to time-sensitive decisions and incident response, LUM provides the Libra Networks--a subsidiary of the Libra Association licensed in Switzerland to operate a payment system--with technical means to respond to critical incidents, as (i) necessitated by its license terms and the law and (ii) as determined by the Libra Network risk appetite/tolerance policy.
```

Reviewer: This is the first introduction of the term LN.

> With respect to complex technical matters, LN can use the TSC, or some other small group, as "advisors" as to what path to take.

Reviewer: I would remove this sentence as it is up to LN's discretion on how to leverage the LUM protocols and this is a technical LIP, not a governance one.

It should be noted that the Association appoints LN officers. Officers are accountable for their actions, and usual measures may be taken upon deviation from their expected conduct.  Hence, the ultimate responsibility and power is by distributed governance of the Association.

> Due to being a reserve backed currency, the true blockchain is the one that is maintained by the Libra Association and reflects the assets in the Libra Reserve. In other words, in contrast to Bitcoin and other “proof of work” and “proof of stake” networks where the miners and users beliefs determine the current fork of the blockchain, with Libra the blockchain used by the reserve determines the true blockchain. Furthermore, Libra Networks operates as a regulated payment company under a Swiss license and must be able to enforce its policy and oversight, as well as to respond to critical operational incidents.

Reviewer: ```suggestion
Due to being a reserve backed currency, the true blockchain is the one that is maintained by the Libra Association and reflects the assets in the Libra Reserve. In other words, in contrast to Bitcoin and other “proof of work” and “proof of stake” networks where the miners and users beliefs determine the current fork of the blockchain, with Libra the blockchain used by the Libra Reserve determines the true blockchain. Furthermore, Libra Networks operates as a regulated payment company under a Swiss license and must be able to enforce its policy and oversight, as well as to respond to critical operational incidents.
```

> The goal of the LUM is to provide the Association with means to exercise its authority to reconfigure the blockchain and to effectively remediate any plausible operational incident while maintaining the trust gained through decentralized governance and distributed systems.

Reviewer: ```suggestion
The goal of LUM is to provide the Association with means to exercise its authority to reconfigure the blockchain and to effectively remediate any plausible operational incident while maintaining the trust gained through decentralized governance and distributed systems.
```

> We first present all building blocks and the reasoning behind, followed by concrete examples.

Reviewer: ```suggestion
We first present all building blocks and the reasoning, followed by concrete examples.
```

# Building blocks
There’s two types of upgrades in the system - updates to the Libra protocol software and changes to on-chain state and we discuss them correspondingly.

## Software upgrade

> Spec changes go through the LIP process and non-spec changes e.g. bug fixes and performance improvements, proceed directly. All software updates are submitted through standard github reviews and must be approved by Maintainers before they are merged to master.

Reviewer: ```suggestion
Spec changes go through the LIP process and non-spec changes (e.g. bug fixes and performance improvements) do not need to initiate the LIP process. All software changes are submitted through standard GitHub reviews and proceed through the code review process before being are merged to master.
```

> Spec changes go through the LIP process and non-spec changes e.g. bug fixes and performance improvements, proceed directly. All software updates are submitted through standard github reviews and must be approved by Maintainers before they are merged to master.

Reviewer: Given this is a formal document, perhaps spec -> specification is better.

> Once an update is ready, the Association needs to release the new version of software and coordinate with all the validator operators the upgrade plan and timeline. Validator operators are responsible for downloading and upgrading the software as per Association instructions.

Reviewer: ```suggestion
Periodically, the Association needs to release a new version of software and coordinate the release plan and timeline with all the validator operators. Validator operators are responsible deploying the software release as per Association instructions.
```

Some spec changes may require all validators to transition at the same height, which is accompanied with an on-chain state value update (described in the next section).

## On-chain state upgrade

> On-chain state is updated automatically via transactions. We propose three ways for the association to execute critical updates, ordered by most frequently expected to least frequently expected: normal multi-sig transactions, multi-sig WriteSet transactions, and genesis transactions.

Reviewer: ```suggestion
On-chain state can be updated via transactions. We propose three ways for the association to execute critical updates, ordered by most frequently expected to least frequently expected: normal multi-sig transactions, multi-sig WriteSet transactions, and genesis transactions.
```

### Multi-sig WriteSet transactions
Writeset transactions are a special type of multi-sig transactions that prescribes an arbitrary state change. Unlike normal transactions, which must follow rules such as “the reserve is the only entity that can create or destroy money”, WriteSet transactions simply describe a state change rather than generate one using the blockchain logic that is encoded in Move. This type of transaction is only needed for extremely rare situations such as updating immutable stdlib libraries due to critical bugs or reversing large scale fraud.

The use of WriteSet transactions follows the spirit of transparency — WriteSet transactions allow an arbitrary change to the state of the blockchain without altering the history of past transactions.
Even though WriteSet transactions can describe arbitrary state change, they are authenticated with the same multi-sig mechanism as normal transactions which means both the change and multi-sig are transparent and visible on-chain.

### Genesis transactions

> Genesis transactions can be either normal or WriteSet transactions. As the name suggests, the process is similar to how the very first genesis transaction works (see steps 3&4 in Starting Libra). The genesis transaction identifies a specific instance of a blockchain run by the Libra protocol. The association is responsible for building the genesis transaction. Users of Libra including validators use the genesis transaction to authenticate the blockchain. If a validator fails to acquire the correct genesis transaction, it will vote on an irrelevant blockchain, but not otherwise impact the system. If a user of the system acquires the incorrect genesis transaction, they will see incorrect state.

Reviewer: We need to make "Starting Libra" public or remove this reference.

Reviewer: association => Association

Everywhere please.

> Note after Libra starts, we may still need genesis transactions as the last resort for disastrous scenarios such as losing more than f validators.

Reviewer: ```suggestion
Note after Libra starts, we may still need genesis transactions as the last resort for extremely unlikely catastrophic scenarios such as losing more than f validators.
```

# Case study

Here's a decision flow chart that summarizes the building blocks and how they will be used in specific examples.

![Decision flow](../static/img/LIP-3-decision-flow.png)

## Network protocol upgrade

> Example: Getting rid of yumux and changing the wire protocol.

Reviewer: Yamux is already gone, so perhaps not a good example.

```suggestion
Example: Changing the wire protocol in a backwards compatible way.
```

Category: Spec change => Software upgrade
## On-chain config update
Example: Adding new validators to the system.

Category: On-chain state change => Multi-sig transaction
## VM upgrade
Example: New bytecode added.

Category: Spec change + require coordination => Software upgrade + Multi-sig transaction
## Consensus protocol upgrade
Example: Switching consensus algorithm from LibraBFTv3 to LibraBFTv4.

Category: Spec change + require coordination => Software upgrade + Multi-sig transaction
## Signature scheme change
Example: Switching consensus signature scheme from EdDSA to BLS.

Category: Spec change + require coordination => Software upgrade + Multi-sig transaction
## Smart contract vulnerability fix/Libra Framework upgrade
Example: DAO hack

Category: On-chain state change + SmartContract incapable => Multi-sig WriteSet transaction
## Large fraud

> Example: Hackers mint billions of libra.

Reviewer: ```suggestion
Example: Hackers mint billions of Libra.
```

Category: On-chain state change + SmartContract incapable => Multi-sig WriteSet transaction
## Quorum Loss

> Example: F+1 validators rage quit the Association

Reviewer: ```suggestion
Example: F+1 validators simultaneously quit the Association and stop running their validators.
```
